### ISSUE TITLE
What's the implementation of the tokenizer?

### ISSUE THREAD

**Issue by kootenpv** [facebookresearch/fastText#133]
And is there any way to tweak it?

**Comment by apiguy** [comment]
Looks like it's pretty simple to me.

https://github.com/facebookresearch/fastText/blob/master/src/dictionary.cc#L147

**Comment by EdouardGrave** [comment]
Hi @kootenpv,

As pointed by @apiguy, the current tokenizer used by fastText is extremely simple: it considers white-spaces as token boundaries. It is thus highly recommended to preprocess the data before feeding it to fastText (e.g. tokenization, lowercasing, etc).

We might add more options for text normalization in the future, but we do not have a timeline yet.

**Comment by sleepdeprivation** [comment]
So the ones in https://github.com/facebookresearch/fastText/blob/master/pretrained-vectors.md are tokenized that way?

I loaded it into a dictionary and get things like `couldnt` and `shouldnt`.

**Comment by qqilihq** [comment]
I agree with @sleepdeprivation -- it would be very helpful to know how the pre-trained data was preprocessed.

**Comment by TechnikEmpire** [comment]
No answer on this? I'm guessing users need to figure out tokenizing languages like Japanese externally / in a preprocessor as suggested here?

**Comment by tranhungnghiep** [comment]
@EdouardGrave In Japanese, there is no whitespace. How does tokenization work in this case? 

Is each symbol a token? If so, there is no subword and most real Japanese words are out-of-vocabulary word in fasttext?

**Comment by tranhungnghiep** [comment]
For future comer, Japanese was "properly" tokenized with Mecab, [source](https://arxiv.org/abs/1802.06893).

**Comment by rth** [comment]
> it would be very helpful to know how the pre-trained data was preprocessed.

According to https://arxiv.org/abs/1802.06893,

> We  used  the  Stanfordword  segmenter  (Chang et al., 2008)  for  Chinese, Mecab  (Kudo, 2005)  for  Japanese  and  UETsegmenter (Nguyen and Le, 2016) for Vietnamese.    For languages written using the Latin, Cyrillic, Hebrew or Greek scripts, we used the tokenizer from the Europarl preprocessing tools (Koehn, 2005).  For the remaining languages, we used the ICU tokenizer.
